v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -0,0 +1,455 @@
+        self._reconnect_daemon = None
+
+        # Event Loop
+        self._loop = asyncio.get_running_loop()
+
+    def _create_mqtt_client(
+        self, client_id, transport, ssl_context, proxy_options, websockets_path
+    ):
+        """
+        Create the MQTT client object and assign all necessary event handler callbacks.
+        """
+        logger.debug("Creating Paho client")
+
+        # Instantiate the client
+        mqtt_client = mqtt.Client(
+            client_id=client_id,
+            clean_session=False,
+            protocol=mqtt.MQTTv311,
+            transport=transport,
+            reconnect_on_failure=False,  # We handle reconnect logic ourselves
+        )
+        if transport == "websockets" and websockets_path:
+            logger.debug("Configuring Paho client for connecting using MQTT over websockets")
+            mqtt_client.ws_set_options(path=websockets_path)
+        else:
+            logger.debug("Configuring Paho client for connecting using MQTT over TCP")
+
+        if proxy_options:
+            logger.debug("Configuring custom proxy options on Paho client")
+            mqtt_client.proxy_set(
+                proxy_type=proxy_options.proxy_type_socks,
+                proxy_addr=proxy_options.proxy_address,
+                proxy_port=proxy_options.proxy_port,
+                proxy_username=proxy_options.proxy_username,
+                proxy_password=proxy_options.proxy_password,
+            )
+
+        mqtt_client.enable_logger(logging.getLogger("paho"))
+
+        # Configure TLS/SSL. If the value passed is None, will use default.
+        mqtt_client.tls_set_context(context=ssl_context)
+
+        # Set event handlers.  Use weak references back into this object to prevent leaks
+        self_weakref = weakref.ref(self)
+
+        def on_connect(client, userdata, flags, rc):
+            this = self_weakref()
+            message = mqtt.connack_string(rc)
+            logger.debug("Connect Response: rc {} - {}".format(rc, message))
+
+            if rc == mqtt.CONNACK_ACCEPTED:
+                # Notify tasks waiting on successful connection
+                async def set_connected():
+                    logger.debug("Client State: CONNECTED")
+                    this._connected = True
+                    this._desire_connection = True
+                    async with this.connected_cond:
+                        this.connected_cond.notify_all()
+
+                f = asyncio.run_coroutine_threadsafe(set_connected(), this._loop)
+                # Need to wait for this one to finish since we don't want to let another
+                # Paho handler invoke until we know the connection state has been set.
+                f.result()
+            else:
+                # Notify tasks waiting on failed connection
+                async def notify_connection_fail():
+                    async with this._connect_failed_cond:
+                        this._connect_failed_cond.notify_all()
```

Comment:
Do we want to include the RC as part of teh notification?

Reply:
simple is good.

Reply:
Updated to use a Future instead of a Condition